notebooks/02-ingestion/04-table-datatypes.ipynb

```diff
@@ -0,0 +1,851 @@
+    "SELECT\n",
+    "  TIME_PARSE(\"depaturetime\") AS \"__time\",\n",
+    "  TIME_PARSE(\"arrivalime\") AS \"__time_arrival\",\n",
+    "  \"Year\",\n",
+    "  \"Reporting_Airline\",\n",
+    "  \"Origin\",\n",
+    "  CASE WHEN \"Origin\" = 'SFO' THEN true ELSE false END AS \"isSilicon\",\n",
+    "  \"Dest\",\n",
+    "  \"Distance\"\n",
+    "FROM \"ext\"\n",
+    "WHERE \"arrivalime\" <> 0\n",
+    "AND \"depaturetime\" <> 0\n",
+    "PARTITIONED BY DAY\n",
+    "'''\n",
+    "\n",
+    "display.run_task(sql)\n",
+    "sql_client.wait_until_ready('example-flights-types-2')\n",
+    "display.table('example-flights-types-2')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f4ef1de0-d230-4969-a3f7-8473ddab48d2",
+   "metadata": {},
+   "source": [
+    "### Timestamps\n",
+    "\n",
+    "Review the SQL statement above used to create the TABLE.\n",
+    "\n",
+    "```sql\n",
+    "  TIME_PARSE(\"depaturetime\") AS \"__time\",\n",
+    "  TIME_PARSE(\"arrivalime\") AS \"__time_arrival\",\n",
+    "```\n",
+    "\n",
+    "The `__time` column has been recognised internally as a true TIMESTAMP, whereas `__time_arrival` is a BIGINT.\n",
+    "\n",
+    "To use `__time_arrival` with timestamp functions and aggregators, it must be converted to a true TIMESTAMP.\n",
+    "\n",
+    "Run the following cell to see two different approaches to this: one using MILLIS_TO_TIMESTAMP function, and one using a CAST."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "80418b6b-fb40-46d7-8eac-879342d33712",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT\n",
+    "  TIME_FLOOR(MILLIS_TO_TIMESTAMP(\"__time_arrival\"),'P1D') AS \"day-1\",\n",
+    "  TIME_FLOOR(CAST(\"__time_arrival\" AS TIMESTAMP),'P1D') AS \"day-2\",\n",
+    "  AVG(TIMESTAMPDIFF(SECOND, \"__time\", NVL(MILLIS_TO_TIMESTAMP(\"__time_arrival\"),\"__time\"))) AS \"averageFlightLength_s\"\n",
+    "FROM \"example-flights-types-2\"\n",
+    "WHERE TIME_IN_INTERVAL(\"__time\",'2005-11-01/P7D')\n",
+    "GROUP BY 1, 2\n",
+    "'''\n",
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c99b46a5-3d31-410a-849f-31f67dc7ec8b",
+   "metadata": {},
+   "source": [
+    "When wanting to parse string data as a timestamp, there are a variety of [functions](https://druid.apache.org/docs/latest/querying/sql-scalar#date-and-time-functions) that support different formats.\n",
```

Comment:
The cell below does not need a FROM clause because it is just processing a constant value. You can use:
```
SELECT CAST('2000-01-02 03:04:05' AS TIMESTAMP) AS "castFromSqlString"
```
The FROM and LIMIT are potentially confusing.

Reply:
Ta - updated this to a nested function that goes to and from timestamps and strings and then a CAST.